src/components/include/cbuf.h

```diff
@@ -141,64 +125,76 @@
- * cbuf that was removed, and 3) before it can return, another thread
- * t_1 enters the cbuf code and finds the stale free-list entry.  The
- * question here is this: How does t_1 know that the free-list entry is
- * stale, and that it should be deallocated?  Answer: If a cbuf is
- * marked as USED when it is seen to be on the free-list, we know this
- * case has been triggered.  
+ * The manager never collect cbufs which are on free-list. Such cbufs
+ * are reachable from client, so they are not garbage. Thus a 
+ * non-null next pointer can prevent a cbuf from being collected.
+ * Whenever the manager removes a cbuf, it will check if it is on a 
+ * free-list (check the "next" pointer in meta data). If it is, the 
+ * manager will set the inconsistent bit in the meta data.
  *
- * Similar to this latter case, when a thread allocates a new cbuf
- * (cbuf_c_create), it should check to see if a freelist descriptor
- * exists before allocating a new one to avoid two descriptors
- * existing for one cbuf.
+ * When a thread gets a cbuf from its free-list, it must check the 
+ * inconsistent bit first. It this is a valid entry then increase the 
+ * refcnt with CAS to detect concurrent modification from the manager.
+ * If CAS fails, it just discards this cbuf and tries to take next 
+ * one. If this is not a valid entry, the client needs to clear the 
+ * inconsistent bit and gets next one.
  *
- * Luckily, the correct way to fix all these inconsistencies is the
- * same: deallocate the allocation free-list descriptor.
+ * Some edge cases:
+ * 1) Before the stale entry is dropped, another thread on different 
+ * core may asks the manager for new cbuf. The manager should avoid to 
+ * allocate the same cbuf id as the invalid entry. To achieve this, 
+ * in cbuf_create the manager always checks if there is invalid meta 
+ * data associated the id in the calling component.
+ * 
+ * 2) a) the manager removes a cbuf with id K from the component c0 (but 
+ * the free-list entry still remains), b) another component c1 creates 
+ * new cbuf. The manager also assigns id K to the new cbuf. c) c1 sends 
+ * this cbuf back to the c0. Does it matter for c0 to receive this cbuf?
+ * No. In cbuf2buf, there is no need to touch free-list and to check 
+ * inconsistent bit. So it can receive this cbuf as normal with those 
+ * flags intact. Later on, when c0 alloc cbuf, it can find the invalid 
+ * entry and skip it. This also implies when discard invalid entry, we 
+ * can only unset the inconsistent bit without modifying other fields.
  */
 
+typedef u32_t cbuf_t; /* fit in a register. */
 typedef union {
 	cbuf_t v;
 	struct {
 		/* 
-		 * cbuf id, aggregate = 1 if this cbuf holds an array of other
-		 * cbufs, tm = 1 if this cbuf is transient (non-persistent)
+		 * cbuf id, aggregate = 1 if this cbuf 
+		 * holds an array of other cbufs
 		 */
-		u32_t aggregate: 1, tm: 1, id:30;
+		u32_t id:31, aggregate: 1;
```

Comment:
this makes a `cbuf_t` for a non-aggregate cbuf identically equal to an unsigned cbid (for x86 / little endian). Though if you treat it as a signed integer the equality is not the same (for cbids > 2^24) I guess. I don't know if this equality is a good thing, as it can permit some sloppiness to use cbuf_t in place of cbid or vice-versa, which will be right "most of the time" now.